pkg/kubelet/kubelet.go

```diff
@@ -745,7 +746,24 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
 	// check node capabilities since the mount path is not the default
 	if len(kubeCfg.ExperimentalMounterPath) != 0 {
 		kubeCfg.ExperimentalCheckNodeCapabilitiesBeforeMount = false
+		resolvePath := filepath.Join(strings.TrimSuffix(kubeCfg.ExperimentalMounterPath, "/mounter"), "rootfs", "etc", "resolv.conf")
+		fileHandle, err := os.OpenFile(resolvePath, os.O_APPEND|os.O_WRONLY, 0600)
+		if err != nil {
+			glog.Errorf("Failed to open file %s, with error %v", resolvePath, err)
+		} else if len(klet.clusterDNS) > 0 {
+			defer fileHandle.Close()
+			dnsString := ""
+			for _, dns := range klet.clusterDNS {
+				dnsString = dnsString + fmt.Sprintf("nameserver %s\n", dns)
+			}
+			dnsString = dnsString + "options rotate\n"
+			glog.V(4).Infof("DNS nameserver string is %s", dnsString)
+			if _, err = fileHandle.WriteString(dnsString); err != nil {
```

Comment:
Just realize we might have problem here. If we are appending kube-dns as an additional nameserver, dns request will be sent to upstream DNS server first. If we query in-cluster DNS name, upstream server will return NXDOMAIN, then the resolver won't check the other server (kube-dns in this case). Ref: https://unix.stackexchange.com/questions/150703/can-subsequent-nameservers-defined-in-etc-resolv-conf-be-used-if-the-previous-n

I think we should overwrite the nameserver entry instead of appending a new one?

Reply:
Does kube-dns forward to upstream dns if it can't resolve the name?

Reply:
Yep it will :)

Reply:
So basically overwrite is enough. If kube-dns could not resolve it, upstream dns will work (just similar to use host dns nameserver)?

Reply:
> So basically overwrite is enough. If kube-dns could not resolve it, upstream dns will work (just similar to use host dns nameserver)?

Correct, this is exactly the way cluster DNS works in k8s :)